### PR TITLE
tls: Add unit test for partial writes

### PR DESCRIPTION
**Pull request by martinpitt**
Set socket buffer of the ws receiving end to a small value and try to
write a large block, so that `connection_write()` runs into a `PARTIAL`
write.

----

I split this out of PR #12461. It's much less intrusive and useful to have at an early point, to make sure this does not regress.

### CODE REVIEW

**Comment by martinpitt** [comment]
Unit tests ran everywhere, so remaining failures are unrelated.

File: src/tls/test-connection.c [hunk 1]
```diff
@@ -150,6 +157,33 @@ test_read_write (void)
 
   g_assert_cmpint (connection_read (c, WS), ==, RETRY);
 
+  /* test partial writes: send a large block */
+  memset (bigmsg, 42, sizeof (bigmsg));
+  g_assert_cmpint (send (client_fds[1], bigmsg, sizeof (bigmsg), 0), ==, sizeof (bigmsg));
+  g_assert_cmpint (connection_read (c, CLIENT), ==, SUCCESS);
+  g_assert_cmpint (c->buf_client.length, ==, sizeof (bigmsg));
+  /* write to ws should be partial due to our SO_SNDBUF from above */
+  g_assert_cmpint (connection_write (c, CLIENT), ==, PARTIAL);
```
Comment (allisonkarlitskaya) [inline src/tls/test-connection.c]: This reads wrong to me, like "aren't we writing to WS, not CLIENT"?  I get the whole thing that it's the data *source*, but... meh.  maybe a comment?

File: src/tls/test-connection.c [hunk 2]
```diff
@@ -150,6 +157,33 @@ test_read_write (void)
...
+    {
+      int len;
+
+      ConnectionResult r = connection_write (c, CLIENT);
+      if (r != SUCCESS && r != RETRY)
+        g_assert_cmpint (r, ==, PARTIAL);
+
+      for (;;)
+        {
+          len = recv (ws_fds[1], buffer, sizeof (buffer), 0);
```
Comment (allisonkarlitskaya) [inline src/tls/test-connection.c]: I could imagine an edge case where connection_write() fails to send data (some imaginary `EINTR` or whatever) and then this `recv()` will block here.  That's never going to happen in this test case, of course....

File: src/tls/test-connection.c [hunk 3]
```diff
@@ -150,6 +157,33 @@ test_read_write (void)
...
+          if (len <= 0)
```
Comment (allisonkarlitskaya) [inline src/tls/test-connection.c]: "On error, terminate the test case, successfully".  Maybe handle the 0 and the -1 cases separately?

File: src/tls/test-connection.c (hunk 2 above)
Comment (martinpitt) [reply src/tls/test-connection.c]: The test does not have any child processes nor alarm()s, etc. Someone could send a SIGUSR1 or something at just the right time, but then again I'd rather have the test fail than introducing a while loop here that hides unexpected signals. I don't have a strong opinion here, other than that I prefer to be as simple and brittle as possible in tests. If you think we need an EINTR loop, I'll add one.

File: src/tls/test-connection.c (hunk 3 above)
Comment (martinpitt) [reply src/tls/test-connection.c]: Ack, done.

File: src/tls/test-connection.c (hunk 1 above)
Comment (martinpitt) [reply src/tls/test-connection.c]: I adjusted the comment.

**Comment by martinpitt** [comment]
Unit tests succeeded everywhere.
